design-proposals/klip-47-non-null-columns-and-component-types.md

```diff
@@ -0,0 +1,183 @@
+```
+
+Declaring a column as `NOT NULL` in a CSAS/CTAS would look like:
+```sql
+CREATE stream pageviews_no_null as select CAST(viewtime AS BIGINT NOT NULL), userid, pageid from pageviews;
+```
+The resulting stream's columns would all be `NOT NULL`
+
+```sql
+ksql> select * from pageviews_no_null emit changes;
++---------------+----------------------------------------------------------------+----------------------------------------------------------------+
+|VIEWTIME       |USERID                                                          |PAGEID                                                          |
++---------------+----------------------------------------------------------------+----------------------------------------------------------------+
+|3              |user                                                            |page                                                            |
+```
+
+Attempting to insert invalid records would error out:
+```sql
+ksql> INSERT INTO pageviews VALUES (5, NULL, 'page23');
+Failed to insert values into 'PAGEVIEWS': null violates the NOT NULL constraint for field "USERID"
+
+ksql> INSERT INTO pageviews VALUES (NULL, 'user3', 'page23'); # this insert succeeds
+
+ksql> INSERT INTO pageviews_no_null VALUES (NULL, 'user3', 'page23');
+Failed to insert values into 'PAGEVIEWS': null violates the NOT NULL constraint for field "VIEWTIME"
+```
+
+`NOT NULL` for arrays, maps, struct
+```sql
+CREATE STREAM valid_data (arr ARRAY<BIGINT NOT NULL>, mapping MAP<VARCHAR NOT NULL, BIGINT NOT NULL>, valid_struct STRUCT<c DOUBLE NOT NULL, d BOOLEAN>) ...
+```
+Records that have column values such as
+* `{arr: [3,4,null] ... }`
+* `{mapping: {"key": null} ...}`
+* `{mapping: {null: 5} ...}`
+* `{valid_struct: {c: null, d: true} ...}`
+
+Would be dropped when reading from the `valid_data` stream and trying to insert record with these column values would error out.
+
+## NOT NULL DEFAULT
+When `ALTER` is used to add a `NOT NULL` constraint in SQL to an existing table column, there can't be any `NULL` values 
+in that column, otherwise there would be rows that violate the `NOT NULL` constraint.
+
+In this case, the user needs to either `UPDATE` all existing `NULL` 
+values in that column or specify a `DEFAULT` value for that column. All existing rows with `NULL` would have the default value added to that column, 
+so that the NOT NULL constraint is not violated (This is only done if the `NOT NULL` constraint is used in conjunction with `DEFAULT`, if a 
+column with no constraints has a `DEFAULT` applied to it, the existing rows with null columns won't be updated).
+
+When `DEFAULT` is specified, instead of dropping the record, the default value would be applied to that record's column.
+```sql
+CREATE stream pageviews_default(viewtime BIGINT, userid VARCHAR NOT NULL DEFAULT("default_user"), pageid VARCHAR NOT NULL DEFAULT("default_page")) ...
```

Comment:
Other DBs use `DEFAULT "default_user"` without parenthesis. Let's use the standard when defining default values.